src/net_processing.cpp

```diff
@@ -254,9 +251,20 @@ class PeerManagerImpl final : public PeerManager
     void SetBestHeight(int height) override { m_best_height = height; };
     void Misbehaving(const NodeId pnode, const int howmuch, const std::string& message) override;
     void ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
-                        const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) override;
+                        const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) override LOCKS_EXCLUDED(m_mutex_message_handling);
 
 private:
+    /** Message handling mutex.
+     *  Message processing is single-threaded, so anything only accessed
+     *  by ProcessMessage(s) or SendMessages can be guarded by this mutex,
+     *  which guarantees it's only accessed by a single thread.
+     */
+    Mutex m_mutex_message_handling;
```

Comment:
If it is single-threaded, then we don't need to guard anything with a mutex?

It looks to me that `PeerManagerImpl::m_mutex_message_handling` is useless because it is only being acquired from a single thread. No two threads are ever going to race to acquire it at the same time:

```cpp
ThreadMessageHandler() // there is just one thread executing this
   ProcessMessages()
       ...
       lock m_mutex_message_handling
       ...
   ...
   SendMessages()
       ...
       lock m_mutex_message_handling
       ...
```

I may be missing something but to me it looks like `m_mutex_message_handling` can be removed safely, which will reduce the complexity of the code.

Reply:
You don't need the mutex at runtime, but having the `GUARDED_BY(m_mutex_message_handling)` at compile time makes it easy to verify nobody's accidentally accessing the object from some other thread (either currently, or as a result of some future change, eg adding an RPC call that wants to expose some of the information stored in those objects). The cost of also having the mutex at runtime is trivial since there's never any contention on it.

Reply:
NACK the addition of `m_mutex_message_handling`.

Following the logic of

> GUARDED_BY() makes it easy to verify nobody's accidentally accessing...

should we add a dummy mutex for **every** variable?

IMO mutexes should be used only where concurrent access is possible.

Reply:
@vasild can you take a look at the commits that build on top of this branch in https://github.com/jnewbery/bitcoin/commits/2021-05-use-mutex-message-handling. Currently, a lot of the data in net_processing is guarded by cs_main, which is also used in validation. One way to loosen that coupling between validation and net_processing is to introduce an internal mutex inside net_processing (`m_mutex_message_handling`) that guards its own members.

Reply:
I think that https://github.com/jnewbery/bitcoin/commits/2021-05-use-mutex-message-handling can be simplified and the common `m_mutex_message_handling` avoided by changing some of the variables to atomics and using a dedicated mutex for e.g. `m_txrequest`.

Btw, it does not compile.
<details>
<summary>minor fixups</summary>

```diff
diff --git i/src/net_processing.cpp w/src/net_processing.cpp
index c56863e5ef..cc04679683 100644
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -401,13 +401,13 @@ private:
      * Set mapBlockSource[hash].second to false if the node should not be
      * punished if the block is invalid.
      */
     std::map<uint256, std::pair<NodeId, bool>> mapBlockSource GUARDED_BY(cs_main);
 
     /** Number of peers with wtxid relay. */
-    int m_wtxid_relay_peers GUARDED_BY(m_mutex_message_handing) {0};
+    int m_wtxid_relay_peers GUARDED_BY(m_mutex_message_handling) {0};
 
     /** Number of outbound peers with m_chain_sync.m_protect. */
     int m_outbound_peers_with_protect_from_disconnect GUARDED_BY(cs_main) = 0;
 
     bool AlreadyHaveTx(const GenTxid& gtxid) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
@@ -1000,20 +1000,15 @@ void UpdateLastBlockAnnounceTime(NodeId node, int64_t time_in_seconds)
     CNodeState *state = State(node);
     if (state) state->m_last_block_announcement = time_in_seconds;
 }
 
 void PeerManagerImpl::InitializeNode(CNode *pnode)
 {
-    LOCK(m_mutex_message_handling);
-
     NodeId nodeid = pnode->GetId();
-    {
-        LOCK(cs_main);
-        mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(pnode->IsInboundConn()));
-        assert(m_txrequest.Count(nodeid) == 0);
-    }
+    WITH_LOCK(cs_main, mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(pnode->IsInboundConn())));
+    WITH_LOCK(m_mutex_message_handling, assert(m_txrequest.Count(nodeid) == 0));
     {
         PeerRef peer = std::make_shared<Peer>(nodeid, m_mutex_message_handling);
         LOCK(m_peer_mutex);
         m_peer_map.emplace_hint(m_peer_map.end(), nodeid, std::move(peer));
     }
     if (!pnode->IsInboundConn()) {
```
</details>

<details>
<summary>compilation errors</summary>

```sh
(clang 13)

net_processing.cpp:986:33: error: reading variable 'fPreferredDownload' requires holding mutex 'peer.m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
    const bool preferred = peer.fPreferredDownload;
                                ^
net_processing.cpp:1033:13: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            _RelayTransaction(txid, tx->GetWitnessHash());
            ^
net_processing.cpp:1062:41: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            nPreferredDownload -= peer->fPreferredDownload;
                                        ^
net_processing.cpp:1063:42: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            m_wtxid_relay_peers -= peer->m_wtxid_relay;
                                         ^
net_processing.cpp:1537:41: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    WITH_LOCK(m_mutex_message_handling, _RelayTransaction(txid, wtxid););
                                        ^
net_processing.cpp:1550:43: error: reading variable 'm_wtxid_relay' requires holding mutex 'it->second->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        node->PushTxInventory(it->second->m_wtxid_relay ? wtxid : txid);
                                          ^
net_processing.cpp:2485:15: error: writing variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' exclusively [-Werror,-Wthread-safety-analysis]
        peer->fPreferredDownload = (!pfrom.IsInboundConn() || pfrom.HasPermission(PF_NOBAN)) && !pfrom.IsAddrFetchConn() && !pfrom.fClient;
              ^
net_processing.cpp:2486:37: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        nPreferredDownload -= peer->fPreferredDownload;
                                    ^
net_processing.cpp:2653:24: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            if (!peer->m_wtxid_relay) {
                       ^
net_processing.cpp:2654:23: error: writing variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' exclusively [-Werror,-Wthread-safety-analysis]
                peer->m_wtxid_relay = true;
                      ^
net_processing.cpp:2777:23: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
            if (peer->m_wtxid_relay) {
                      ^
net_processing.cpp:3049:37: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        const uint256& hash = peer->m_wtxid_relay ? wtxid : txid;
                                    ^
net_processing.cpp:3051:19: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        if (peer->m_wtxid_relay && txid != wtxid) {
                  ^
net_processing.cpp:3075:13: error: calling function 'AlreadyHaveTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        if (AlreadyHaveTx(GenTxid(/* is_wtxid=*/true, wtxid))) {
            ^
net_processing.cpp:3084:21: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
                    _RelayTransaction(tx.GetHash(), tx.GetWitnessHash());
                    ^
net_processing.cpp:3090:44: error: calling function 'AcceptToMemoryPool' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
        const MempoolAcceptResult result = AcceptToMemoryPool(m_chainman.ActiveChainstate(), m_mempool, ptx, false /* bypass_limits */);
                                           ^
net_processing.cpp:3094:23: error: calling function 'check' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            m_mempool.check(m_chainman.ActiveChainstate());
                      ^
net_processing.cpp:3099:13: error: calling function '_RelayTransaction' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            _RelayTransaction(tx.GetHash(), tx.GetWitnessHash());
            ^
net_processing.cpp:3114:13: error: calling function 'ProcessOrphanTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
            ProcessOrphanTx(*peer);
            ^
net_processing.cpp:3131:21: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                if (recentRejects->contains(parent_txid)) {
                    ^
net_processing.cpp:3147:26: error: calling function 'AlreadyHaveTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
                    if (!AlreadyHaveTx(gtxid)) AddTxAnnouncement(pfrom, *peer, gtxid, current_time);
                         ^
net_processing.cpp:3172:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetHash());
                ^
net_processing.cpp:3173:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetWitnessHash());
                ^
net_processing.cpp:3192:24: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                assert(recentRejects);
                       ^
net_processing.cpp:3193:17: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                recentRejects->insert(tx.GetWitnessHash());
                ^
net_processing.cpp:3204:21: error: reading variable 'recentRejects' requires holding mutex 'cs_main' [-Werror,-Wthread-safety-analysis]
                    recentRejects->insert(tx.GetHash());
                    ^
net_processing.cpp:4291:29: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
        bool fFetch = peer->fPreferredDownload || (nPreferredDownload == 0 && !pto->fClient && !pto->IsAddrFetchConn()); // Download if this is a nice peer, or we have no nice peers and this one might do.
                            ^
net_processing.cpp:4504:53: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        const uint256& hash = peer->m_wtxid_relay ? txinfo.tx->GetWitnessHash() : txinfo.tx->GetHash();
                                                    ^
net_processing.cpp:4505:40: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
                                       ^
net_processing.cpp:4536:85: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                    CompareInvMempoolOrder compareInvMempoolOrder(&m_mempool, peer->m_wtxid_relay);
                                                                                    ^
net_processing.cpp:4548:40: error: reading variable 'm_wtxid_relay' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                        CInv inv(peer->m_wtxid_relay ? MSG_WTX : MSG_TX, hash);
                                       ^
net_processing.cpp:4636:117: error: reading variable 'fPreferredDownload' requires holding mutex 'peer->m_mutex_message_handling' [-Werror,-Wthread-safety-analysis]
                if (current_time > state.m_headers_sync_timeout && nSyncStarted == 1 && (nPreferredDownload - peer->fPreferredDownload >= 1)) {
                                                                                                                    ^
32 errors generated.
```
</details>

Reply:
"Should we add a dummy mutex for every variable" -- we should have a GUARDED_BY for every object that's reachable by multiple threads, yes. The mutex is not a dummy, it's a real mutex.

For this specific case, here's an example of a patch that adds an access from another thread: https://github.com/ajtowns/bitcoin/commits/202105-whyhaveaguard . With the `GUARDED_BY(m_mutex_message_handling)` this is caught at compile time; without the guard (removed in the second commit), it compiles fine but introduces a race condition.

Reply:
> we should have a GUARDED_BY for every object that's reachable by multiple threads

I assume here you mean -- even if the current code accesses it from a single thread.

If yes, then I think that is an overkill - it means having a mutex attached to every non-const global or class member variable, including private ones (coz in the future somebody may access it from another thread).

> For this specific case, here's an example...

Yeah, but one can do such an example for any other variable. IMO annotating everything with `GUARDED_BY()`, just in case, would have a net-adverse effect on the code base.

<details>
<summary>Similar example for two other, randomly picked variables</summary>

```diff
diff --git i/src/net.h w/src/net.h
index 6b9a7ef136..18d57a7d8e 100644
--- i/src/net.h
+++ w/src/net.h
@@ -1256,12 +1256,18 @@ private:
      * an address and port that are designated for incoming Tor connections.
      */
     std::vector<CService> m_onion_binds;
 
     friend struct CConnmanTest;
     friend struct ConnmanTestMsg;
+
+public:
+    size_t ResponseCachesSize() const
+    {
+        return m_addr_response_caches.size();
+    }
 };
 
 /** Return a timestamp in the future (in microseconds) for exponentially distributed events. */
 std::chrono::microseconds PoissonNextSend(std::chrono::microseconds now, std::chrono::seconds average_interval);
 
 /** Dump binary message to file, with timestamp */
diff --git i/src/rpc/net.cpp w/src/rpc/net.cpp
index 1f6b6e8d7e..cfab1ebafc 100644
--- i/src/rpc/net.cpp
+++ w/src/rpc/net.cpp
@@ -631,12 +631,14 @@ static RPCHelpMan getnetworkinfo()
     NodeContext& node = EnsureAnyNodeContext(request.context);
     if (node.connman) {
         ServiceFlags services = node.connman->GetLocalServices();
         obj.pushKV("localservices", strprintf("%016x", services));
         obj.pushKV("localservicesnames", GetServicesNames(services));
     }
+    obj.pushKV("bug1", node.connman->ResponseCachesSize());
+    obj.pushKV("bug2", node.addrman->m_asmap.size());
     if (node.peerman) {
         obj.pushKV("localrelay", !node.peerman->IgnoresIncomingTxs());
     }
     obj.pushKV("timeoffset",    GetTimeOffset());
     if (node.connman) {
         obj.pushKV("networkactive", node.connman->GetNetworkActive());
```
</details>

Reply:
> `diff --git i/src/net.h w/src/net.h`

Yes, net.cpp/net.h is terrible for having globally accessible variables that aren't annotated for thread safety. Being able to introduce races that the compiler doesn't catch is a bad thing... txmempool's a better example, though it's also missing some guards (cf #22003).

> IMO annotating everything with `GUARDED_BY()`, just in case, would have a net-adverse effect on the code base.

`GUARDED_BY` is always a "just in case" thing -- any time the code compiles correctly with `GUARDED_BY` it will also compile correctly without it -- the benefit is "just in case" someone forgets while modifying the code later. With a guard, the compiler will catch the mistake; without it, you have to hope someone reviewing the code remembers, or wait for some obscure bug to be noticed and reported.

Not consistently guarding variables has an easily observable adverse affect on the code bas as a whole: it makes it hard to change code, because you're not sure when things might be safe to reuse in a different context. eg see https://github.com/bitcoin/bitcoin/pull/21061#discussion_r602283867

>  it means having a mutex attached to every non-const global or class member variable, including private ones

Yes and no -- no, in that most classes don't need to manually manage locks, but should rather rely on whatever instantiates them to ensure the entire object is accessed safely; but for the ones that do, yes: every non-const, non-atomic member should be guarded (as should every class's non-const, non-atomic static members, since they're effectively globals themselves).

Reply:
@ajtowns, thanks for the elaboration.

I still think that mutexes should be introduced when needed, not beforehand (with a reason that in the future somebody may access a variable from a different thread).

Looks like neither one of us is convinced by the other's arguments. It's ok, it would have been too boring if everybody agreed on everything all the time.

Reply:
I think it is fine to not have thread safety stuff on things that aren't meant to be called by more than one thread. For example, most code in init.cpp doesn't need them because they are only called once per process. Though, other stuff should communicate to the developer whether it is thread safe. Either via code comments (https://github.com/bitcoin/bitcoin/blob/e3de7cb9039770e0fd5b8bb8a5cba35c87ae8f00/src/random.h#L67, https://github.com/bitcoin/bitcoin/blob/e3de7cb9039770e0fd5b8bb8a5cba35c87ae8f00/src/random.h#L129) or via annotations.